src/Form.php

```diff
@@ -323,30 +314,53 @@ public function _fieldFactory(\atk4\data\Field $f)
                 $display = [$display];
             }
 
-            // ui['form'] = ['FormField/TextArea', 'rows'=>2]
             if (isset($display[0])) {
-                $display = array_merge($display, $arg);
-
-                return $this->factory($display);
+                $class = $class ?: $display[0];
+                unset($display[0]);
             }
+
+            $defaults = array_merge($display, $defaults);
         }
 
-        if ($f->enum) {
-            $arg['values'] = array_combine($f->enum, $f->enum);
+        if (!$class && $f->enum) {
+            $defaults['values'] = array_combine($f->enum, $f->enum);
 
-            return new FormField\Dropdown($arg);
+            $class = 'Dropdown';
         }
 
         // Field values can be picked from the model.
         if (isset($f->reference)) {
-            //$arg['values'] = $f->ref();
+            $class = 'Dropdown';
+            $defaults['model'] = $f->reference->refModel();
+            //$dd = new FormField\Dropdown($arg);
+            //$dd->setModel($f->reference->refModel());
+            //return $dd;
+        }
 
-            $dd = new FormField\Dropdown($arg);
-            $dd->setModel($f->reference->refModel());
+        if (isset($this->typeToDecorator[$f->type])) {
+            $class = $this->typeToDecorator[$f->type];
+        }
 
-            return $dd;
+        if (!$class) {
+            $class = 'Line';
         }
 
+        return $this->factory($class, $defaults, 'FormField');
+    }
+
+    protected $typeToDecorator = [
+        'boolean' => 'Checkbox',
+        'text'    => 'Textarea',
+        'string'  => 'Line',
+        'password'=> 'Password',
+        'datetime'=> 'Datetime',
+        'date'    => ['Calendar', 'type'=>'date'],
```

Comment:
there are no Datetime, Calendar and Time decorators in repo

Reply:
Calendar is there, others I have changed.